### PR TITLE
Simplify Snapshot Delete Process

### PR DESCRIPTION
**Pull request by original-brownbear**
We don't need to read the SnapshotInfo for a
snapshot to determine the indices that need to
be updated when it is deleted as the `RepositoryData`
contains that information already.
This PR makes it so the `RepositoryData` is used to
determine which indices to update and also removes
the special handling for deleting snapshot metadata
and the CS snapshot blob and has those simply be
deleted as part of the deleting of other unreferenced
blobs in the last step of the delete.

This makes the snapshot delete a little faster and
more resilient by removing two RPC calls
(the separate delete and the get).

Also, this shortens the diff with #46250 as a
side-effect.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -365,6 +359,8 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
             try {
                 final Map<String, BlobMetaData> rootBlobs = blobContainer().listBlobs();
                 final RepositoryData repositoryData = getRepositoryData(latestGeneration(rootBlobs.keySet()));
+                // Cache the indices that were found before writing out the new index-N blob so that a stuck master will never
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This comment belongs here, it must have accidentally gotten out of place in a previous refactoring.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by original-brownbear** [comment]
Jenkins update branch

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 2]
```diff
@@ -123,6 +123,20 @@ public SnapshotState getSnapshotState(final SnapshotId snapshotId) {
...
     }
 
+    /**
+     * Returns the list of {@link IndexId} that have their snapshots updated but not removed (because they are still referenced by other
+     * snapshots) after removing the given snapshot from the repository.
+     *
+     * @param snapshotId SnapshotId to remove
+     * @return List of indices that are changed but not removed
+     */
+    public List<IndexId> indicesToUpdateAfterRemovingSnapshot(SnapshotId snapshotId) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: Can you add a unit test for this in RepositoryDataTests?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: How about https://github.com/elastic/elasticsearch/pull/47439/commits/0961fd1062162bcb6578162bb9e36c4155555705 ?

(it's in the style of the other tests for this class, could also hard code a simpler case if you want :))

**Review by tlrx — APPROVED** [review]
LGTM, thanks Armin!

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java (hunk 2 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: It's good, thanks

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**